_posts/2018-11-08-google-cpp-style-guide-is-no-good.md

```diff
@@ -0,0 +1,297 @@
+
+I think this rule mixes together two ideas for no good reason. C++ has a clear way to denote mutability -- the `const` qualifier.
+Using pointer vs reference to signal mutability goes against the language.
+It makes more sense to use a pointer when `nullptr` is allowed and a reference otherwise.
+This would eliminate a runtime assertion for unexpected null pointer in the body of the function.
+
+The reasoning behind this rule is to establish a convention that communicates mutability at the call site via the presence of the address-of operator.
+In other words, the reader is expected to understand the semantics of an unknown function without consulting the declaration (or documentation).
+Maybe it works for understanding mutability of arguments but I doubt the readers will get very far with just this knowledge.
+
+### [Exceptions](https://google.github.io/styleguide/cppguide.html#Exceptions)
+Exceptions vs error codes debate is much like space-vs-tabs so I will sit this one out.
+GSG forbids exceptions on the ground of "Google's existing code is not exception-tolerant".
+I'm not sure what that refers to.
+Did they mark every function `noexcept` or are they using `goto fail` convention for cleanup?
+
+What I find remarkable is that it offers no guideance on how to deal with errors in the absence of exceptions.
+* Should developers use `errno` global?
+* Should they use something like the proposed `std::expected`?
+* Should every function return an error code and use out-parameters to return computed values?
+
+### [Lambda expressions](https://google.github.io/styleguide/cppguide.html#Exceptions)
+"Use lambda expressions where appropriate." -- probably a cheap shot, but what is the alternative? -- use them where it's not appropriate?
+
+### [Template metaprogramming](https://google.github.io/styleguide/cppguide.html#Template_metaprogramming)
+For the most part, GSG tries to scare anyone away from TMP.
+Unfortunately Google is not alone here as TMP is often frowned upon in many C++ circles.
+I have heard many opponents of TMP and their objections are almost always grounded in the fear of someone just using it to be clever.
+
+It seems that many engineers simply don't understand what metaprogramming is and what it's used for.
+When one needs to generate code (write a meta-program), it can either be done from within C++ (via TMP) or using an external script (e.g. written in Python) or a separate compiler (e.g. lex, yacc, swig).
+I will assert that using an external script/compiler ends up being worse for a number of reasons:
+* You need to integrate the tool into your toolchain
+* You need to keep the data-structures in sync
+* Crossing the boundary often requires going through a C interface
+* Having scripts that output C++ code are not easy to develop or maintain
+
+The use of a separate DSL does have advantages and can be justified if it will be a large part of the project.
```

Comment:
not sure you defined what DSL means prior to using the abbreviation

Reply:
Good catch